README
NAME
    Data::Semantic - Common API for data with semantics attached to them

VERSION
    version 1.100820

DESCRIPTION
    There are various distributions on CPAN dealing with data that has
    semantics attached them. For example, an IP address is not just a
    string, but it has a format the can be checked.

    There is Data::Domain. It does not store the values to be tested, it
    just tests them and records messages.

    Then there is Class::Value. It does store the value and it records
    exceptions.

    Both share the same underlying concept of a value that can be tested for
    validity. Data::Semantic provides this underlying mechanism. It does not
    store the value and it does not record the message. All it does is
    validate data and say whether it is good or bad. What Data::Domain or
    Class::Value do with this information is their thing.

    If you have a new type of semantic data, here is how you can handle it.

        if it can be a Regexp::Common pattern:
            make a Regexp::Common pattern
            make a subclass of Data::Semantic::RegexpAdapter
        else:
            make a subclass of Data::Semantic::Base
            if there is a CPAN module for that data type:
                wrap that module in your semantic data class
                forward methods onto the object if possible

        make an adapter class for Class::Value
        make an adapter class for Data::Domain

METHODS
  new
        my $obj = Data::Semantic->new;
        my $obj = Data::Semantic->new(%args);

    Creates and returns a new object. The constructor will accept as
    arguments a list of pairs, from component name to initial value. For
    each pair, the named component is initialized by calling the method of
    the same name with the given value. If called with a single hash
    reference, it is dereferenced and its key/value pairs are set as
    described before.

  is_valid
    FIXME

  is_valid_normalized_value
    FIXME

  normalize
    FIXME

INSTALLATION
    See perlmodinstall for information and options on installing Perl
    modules.

BUGS AND LIMITATIONS
    No bugs have been reported.

    Please report any bugs or feature requests through the web interface at
    <http://rt.cpan.org/Public/Dist/Display.html?Name=Data-Semantic>.

AVAILABILITY
    The latest version of this module is available from the Comprehensive
    Perl Archive Network (CPAN). Visit <http://www.perl.com/CPAN/> to find a
    CPAN site near you, or see <http://search.cpan.org/dist/Data-Semantic/>.

    The development version lives at
    <http://github.com/hanekomu/Data-Semantic/>. Instead of sending patches,
    please fork this project using the standard git and github
    infrastructure.

AUTHOR
      Marcel Gruenauer <marcel@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2007 by Marcel Gruenauer.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Perl extension Data-Semantic

1.100820  2010-03-23 17:14:39 Europe/Vienna
    - converted the distribution to Dist::Zilla-style

0.04  Thu 2010.02.11 15:03:36 CET (Marcel Gruenauer <marcel@cpan.org>)
    - Changed dist style and Changes back to standard. People didn't like it -
      the nail that sticks out gets hammered down.
    - Added standard test files; this will also help with CPANTS.

0.03  Thu, 29 May 2008 12:03:26 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - set the version to 0.03
    - moved generated documentation of interited methods to after documentation
      of this class' methods so the manpage is more readable
    - .shipit: fixed svk.tagpattern

0.02  Thu, 22 Nov 2007 13:08:38 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - set the version to 0.02
    - Makefile.PL: now requires Test::CompanionClasses 0.03; earlier versions
      had a bug in make_real_object()

0.01  Wed, 21 Nov 2007 18:22:03 -0000 (Marcel Gruenauer <marcel@cpan.org>)
    - original version
    - Data::Semantic::Base: is_valid_normalized_value() also requires the value
      to be well-formed
    - Data::Semantic::RegexpAdapter: is_valid_normalized_value() just calls
      is_well_formed() - is that right?
    - Data::Semantic::Test: now deals with TESTDATA instead of well_formed(),
      not_well_formed(), valid() and invalid()
    - updated MANIFEST
    - deleted lib/Data/Semantic/RegexpTest.pm
    - getting rid of the distinction between well-formedness and validity.
      Only validity remains
    - deleted lib/Data/Semantic.pm
    - moveed Data::Semantic::Base to Data::Semantic
    - deleted lib/Data/Semantic/Base.pm
    - started to add documentation
    - Data::Semantic::RegexpAddapter: keep() is now a boolean attribute
    - Data::Semantic::Test: added documentation
    - updated t/perlcriticrc


lib/Data/Semantic/Test.pm
use 5.008;
use strict;
use warnings;

package Data::Semantic::Test;
our $VERSION = '1.100820';
# ABSTRACT: Testing Data::Semantic objects
use Test::More;
use parent 'Test::CompanionClasses::Base';

sub PLAN {
    my $self = shift;
    my $plan = 0;
    for my $test ($self->TESTDATA) {

        # see run() about the multipliers
        $plan += @{ $test->{valid} || [] } + @{ $test->{invalid} || [] };
    }
    $plan;
}

sub munge_args {
    my ($self, %args) = @_;
    %args;
}

sub test_is_valid {
    my ($self, $obj, $value, $testname) = @_;
    ok($obj->is_valid($value), $testname);
}

sub test_is_invalid {
    my ($self, $obj, $value, $testname) = @_;
    ok(!$obj->is_valid($value), $testname);
}

sub run {
    my $self = shift;
    $self->SUPER::run(@_);
    for my $test ($self->TESTDATA) {
        my %args = $self->munge_args(%{ $test->{args} || {} });

        # string representation for test name
        my $args = join ',' => map { "$_=$args{$_}" }
          sort keys %args;
        $args = '(none)' unless length $args;
        my $obj = $self->make_real_object(%args);
        $self->test_is_valid($obj, $_, "VALID   $args: $_")
          for @{ $test->{valid} || [] };

        # If a value is not even well-formed, it most certainly is not valid,
        # so add these tests as well.
        $self->test_is_invalid($obj, $_, "INVALID $args: $_")
          for @{ $test->{invalid} || [] };
    }
}
1;


__END__
=pod

=head1 NAME

Data::Semantic::Test - Testing Data::Semantic objects

=head1 VERSION

version 1.100820

=head1 DESCRIPTION

This class makes it easy to test new semantic data classes based on
L<Data::Semantic>. It uses the L<Test::CompanionClasses> mechanism. So to
test the subclass L<Data::Semantic::URI::http> you would write a corresponding
L<Data::Semantic::URI::http_TEST> test class. In your test class you need to
define the following structure:

    use constant TESTDATA => (
        {
            args => {},
            valid => [ qw(
                http://localhost/
                http://use.perl.org/~hanekomu/journal?entry=12345
            ) ],
            invalid  => [ qw(
                news://localhost/
                http://?123
                https://localhost/
            ) ],
        },
        {
            args => { scheme => 'https?' },
            valid => [ qw(
                http://localhost/
                http://use.perl.org/~hanekomu/journal?entry=12345
                https://localhost/
                https://use.perl.org/~hanekomu/journal?entry=12345
            ) ],
            invalid  => [ qw(
                news://localhost/
                http://?123
            ) ],
        },
        {
            args => { scheme => 'https' },
            valid => [ qw(
                https://localhost/
                https://use.perl.org/~hanekomu/journal?entry=12345
            ) ],
            invalid  => [ qw(
                http://localhost/
                http://use.perl.org/~hanekomu/journal?entry=12345
                http://?123
                news://localhost/
            ) ],
        },
    );

So you define one or more scenarios, each within its own hashref within the
C<TESTDATA> list. In each scenario you have a list of arguments to pass to the
semantic data object constructor. Given those arguments, certain values will
be considered valid and others invalid. 

See L<Test::CompanionClasses> for more information on how these tests are run.

=head1 METHODS

=head2 PLAN

FIXME

=head2 run

FIXME

=head2 munge_args

FIXME

=head2 test_is_invalid

FIXME

=head2 test_is_valid

FIXME

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Data-Semantic>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Data-Semantic/>.

The development version lives at
L<http://github.com/hanekomu/Data-Semantic/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Data/Semantic.pm
use 5.008;
use strict;
use warnings;

package Data::Semantic;
our $VERSION = '1.100820';
# ABSTRACT: Common API for data with semantics attached to them
use parent qw(Class::Accessor::Complex);
__PACKAGE__->mk_new;

sub is_valid {
    my ($self, $value) = @_;

    # value can be undef
    return 1 unless defined $value;
    my $normalized = $self->normalize($value);
    return 0 unless defined $normalized;
    $self->is_valid_normalized_value($normalized);
}

sub normalize {
    my ($self, $value) = @_;
    $value;
}

sub is_valid_normalized_value {
    my ($self, $normalized) = @_;
    defined $normalized;
}
1;


__END__
=pod

=head1 NAME

Data::Semantic - Common API for data with semantics attached to them

=head1 VERSION

version 1.100820

=head1 DESCRIPTION

There are various distributions on CPAN dealing with data that has semantics
attached them. For example, an IP address is not just a string, but it has a
format the can be checked.

There is L<Data::Domain>. It does not store the values to be tested, it just
tests them and records messages.

Then there is L<Class::Value>. It does store the value and it records
exceptions.

Both share the same underlying concept of a value that can be tested for
validity. Data::Semantic provides this underlying mechanism. It does not store
the value and it does not record the message. All it does is validate data and
say whether it is good or bad. What L<Data::Domain> or L<Class::Value> do with
this information is their thing.

If you have a new type of semantic data, here is how you can handle it.

    if it can be a Regexp::Common pattern:
        make a Regexp::Common pattern
        make a subclass of Data::Semantic::RegexpAdapter
    else:
        make a subclass of Data::Semantic::Base
        if there is a CPAN module for that data type:
            wrap that module in your semantic data class
            forward methods onto the object if possible

    make an adapter class for Class::Value
    make an adapter class for Data::Domain

=head1 METHODS

=head2 new

    my $obj = Data::Semantic->new;
    my $obj = Data::Semantic->new(%args);

Creates and returns a new object. The constructor will accept as arguments a
list of pairs, from component name to initial value. For each pair, the named
component is initialized by calling the method of the same name with the given
value. If called with a single hash reference, it is dereferenced and its
key/value pairs are set as described before.

=head2 is_valid

FIXME

=head2 is_valid_normalized_value

FIXME

=head2 normalize

FIXME

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Data-Semantic>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Data-Semantic/>.

The development version lives at
L<http://github.com/hanekomu/Data-Semantic/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



lib/Data/Semantic/RegexpAdapter.pm
use 5.008;
use strict;
use warnings;

package Data::Semantic::RegexpAdapter;
our $VERSION = '1.100820';
# ABSTRACT: Adapter for Regexp::Common patterns
use Regexp::Common;
use parent qw(
  Data::Semantic
  Data::Inherited
);
#<<<
__PACKAGE__
    ->mk_scalar_accessors(qw(re))
    ->mk_boolean_accessors(qw(keep))
    ->mk_hash_accessors(qw(kept));
#>>>
use constant LOAD        => '';
use constant REGEXP_KEYS => ();
use constant KEEP_KEYS   => ();

sub init {
    my $self = shift;
    Regexp::Common->import($self->LOAD) if $self->LOAD;
    my @regexp_keys = $self->every_list('REGEXP_KEYS');
    @regexp_keys || die "REGEXP_KEYS is not defined";
    my $re_spec = sprintf '$RE%s', join '' => map { "{$_}" } @regexp_keys,
      $self->flags;
    my $re = eval $re_spec;
    die $@ if $@;
    $self->re($re);
}

# turn the object's settings into a list of flags to be passed to
# Regexp::Common's $RE
sub flags {
    my $self = shift;
    my @flags;
    push @flags => '-keep' if $self->keep;
    @flags;
}

sub is_valid_normalized_value {
    my ($self, $value) = @_;
    my $re = $self->re;
    my $ok = $value =~ /^$re$/;

    # assume {-keep} was given in any case - the user will know whether kept()
    # will return anything useful.
    my %keep;
    @keep{ $self->KEEP_KEYS } =
      map { defined($_) ? $_ : '' } ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10);
    $self->kept(%keep);
    $ok;
}
1;


__END__
=pod

=head1 NAME

Data::Semantic::RegexpAdapter - Adapter for Regexp::Common patterns

=head1 VERSION

version 1.100820

=head1 METHODS

=head2 LOAD

    use constant LOAD => 'AT::NICAT';

This is the optional name of the Regexp::Common module to load. For example,
if you use a pattern from L<Regexp::Common::AT::NICAT>, you would set this
to C<AT::NICAT>. If you use patterns bundled in the same distribution as
Regexp::Common you can leave it empty.

=head2 REGEXP_KEYS

    use constant REGEXP_KEYS => qw(URI file);

These is the list of keys that you would pass to Regexp::Common's C<$RE>. For
example, if you wanted to match HTTP URIs, you would use C<qw(URI HTTP)>.
Compare with L<Regexp::Common::URI::http>. See L<Regexp::Common> for more
details on this mechanism.

=head2 KEEP_KEYS

    use constant KEEP_KEYS => qw(scheme host port query);

This class supports Regexp::Common's C<-keep> mechanism. C<kept()> returns a
hash of the patterns returned by Regexp::Common. In this list you can specify
the hash keys that C<$1>, C<$2> and so on are mapped to.

=head2 flags

Turns the object's settings into a list of flags to be passed to
Regexp::Common's C<$RE>. For example, Regexp::Common expects a C<{-keep}> key,
but this class has a C<keep()> accessor. If you subclass this class and
add more accessors that correspond to Regexp::Common keys, you need to
override this method and map the attributes to the keys. Be sure to call
C<SUPER::flags()>. See L<Data::Semantic::URI::http> for an example.

=head2 init

FIXME

=head2 is_valid_normalized_value

FIXME

=head1 INSTALLATION

See perlmodinstall for information and options on installing Perl modules.

=head1 BUGS AND LIMITATIONS

No bugs have been reported.

Please report any bugs or feature requests through the web interface at
L<http://rt.cpan.org/Public/Dist/Display.html?Name=Data-Semantic>.

=head1 AVAILABILITY

The latest version of this module is available from the Comprehensive Perl
Archive Network (CPAN). Visit L<http://www.perl.com/CPAN/> to find a CPAN
site near you, or see
L<http://search.cpan.org/dist/Data-Semantic/>.

The development version lives at
L<http://github.com/hanekomu/Data-Semantic/>.
Instead of sending patches, please fork this project using the standard git
and github infrastructure.

=head1 AUTHOR

  Marcel Gruenauer <marcel@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Marcel Gruenauer.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut



Makefile.PL

use strict;
use warnings;

 BEGIN { require 5.008; } 

use ExtUtils::MakeMaker 6.11;



my %WriteMakefileArgs = (
                       'test' => {
                                   'TESTS' => 't/*.t'
                                 },
                       'NAME' => 'Data::Semantic',
                       'DISTNAME' => 'Data-Semantic',
                       'CONFIGURE_REQUIRES' => {
                                                 'ExtUtils::MakeMaker' => '6.11'
                                               },
                       'AUTHOR' => 'Marcel Gruenauer <marcel@cpan.org>',
                       'BUILD_REQUIRES' => {},
                       'ABSTRACT' => 'Common API for data with semantics attached to them',
                       'EXE_FILES' => [],
                       'VERSION' => '1.100820',
                       'PREREQ_PM' => {
                                        'Class::Accessor::Complex' => '0',
                                        'Scalar::Util' => '0',
                                        'parent' => '0',
                                        'English' => '0',
                                        'constant' => '0',
                                        'Carp' => '0',
                                        'Test::More' => '0.88',
                                        'Test::CompanionClasses::Base' => '0',
                                        'Regexp::Common' => '0',
                                        'File::Find' => '0',
                                        'File::Temp' => '0',
                                        'Data::Inherited' => '0'
                                      },
                       'LICENSE' => 'perl'
                     );


delete $WriteMakefileArgs{LICENSE}
  unless eval { ExtUtils::MakeMaker->VERSION(6.31) };

WriteMakefile(%WriteMakefileArgs);



